fix Rust panic error in signature creation

The following error was triggered by this command line, in #1159 -
```
sourmash sketch protein -p k=21,dna tests/test-data/ecoli.faa -o foo.sig
```

error:
```
== This is sourmash version 3.5.0a0. ==
== Please cite Brown and Irber (2016), doi:10.21105/joss.00027. ==

computing signatures for files: tests/test-data/ecoli.faa
Computing a total of 1 signature(s).
... reading sequences from tests/test-data/ecoli.faa
Traceback (most recent call last):
  File "/Users/t/miniconda3/envs/py37/bin/sourmash", line 11, in <module>
    load_entry_point('sourmash', 'console_scripts', 'sourmash')()
  File "/Users/t/dev/sourmash/sourmash/__main__.py", line 13, in main
    return mainmethod(args)
  File "/Users/t/dev/sourmash/sourmash/cli/sketch/protein.py", line 69, in main
    return sourmash.command_sketch.protein(args)
  File "/Users/t/dev/sourmash/sourmash/command_sketch.py", line 220, in protein
    _execute_sketch(args, signatures_factory)
  File "/Users/t/dev/sourmash/sourmash/command_sketch.py", line 175, in _execute_sketch
    _compute_individual(args, signatures_factory)
  File "/Users/t/dev/sourmash/sourmash/command_compute.py", line 228, in _compute_individual
    args.input_is_protein, args.check_sequence)
  File "/Users/t/dev/sourmash/sourmash/command_compute.py", line 281, in add_seq
    sig.add_protein(seq)
  File "/Users/t/dev/sourmash/sourmash/signature.py", line 155, in add_protein
    self._methodcall(lib.signature_add_protein, to_bytes(sequence))
  File "/Users/t/dev/sourmash/sourmash/utils.py", line 25, in _methodcall
    return rustcall(func, self._get_objptr(), *args)
  File "/Users/t/dev/sourmash/sourmash/utils.py", line 78, in rustcall
    raise exc
sourmash.exceptions.Panic: sourmash panicked: thread 'unnamed' panicked with 'called `Result::unwrap()` on an `Err` value: InvalidHashFunction { function: "dna" }' at src/core/src/signature.rs:367
```

there is now a check in the CLI code to prevent this, and a test, but that can be bypassed by commenting out the `Incompatible sketch type` ValueError exceptions in `command_sketch.py` if we want to reproduce this.

see @luizirber comment [here](https://github.com/dib-lab/sourmash/pull/1159#issuecomment-673668188)

I said:
> throws an error already :). I'll have to make this a nicer error message, is all!
> `sourmash.exceptions.Panic: sourmash panicked: thread 'unnamed' panicked with 'called `Result::unwrap()` on an `Err` value: InvalidHashFunction { function: "dna" }' at src/core/src/signature.rs:367`

luiz said:

Yikes, a `Panic`... Oh, I see, this is the code where it is being thrown:
```rust
self.signatures 
    .iter_mut()  
    .for_each(|sketch| { 
        sketch.add_protein(&seq).unwrap(); }  
    );
```
I didn't know about `try_for_each` before, which allows errors, so the fix is
```rust
self.signatures 
    .iter_mut()  
    .try_for_each(|sketch| { 
        sketch.add_protein(&seq) }  
    )?;
```
This is going to raise a `ValueError`, which you can then capture and give a better error message

----

I made the changes in src/core/src/signature.rs, line 355 / method `add_protein`, and got:

```
   Compiling sourmash v0.9.0 (/Users/t/dev/sourmash/src/core)
error[E0277]: the trait bound `(): std::ops::Try` is not satisfied
   --> src/core/src/signature.rs:366:18
    |
366 |                 .try_for_each(|sketch| {
    |                  ^^^^^^^^^^^^ the trait `std::ops::Try` is not implemented for `()`

error: aborting due to previous error

For more information about this error, try `rustc --explain E0277`.
error: could not compile `sourmash`.

To learn more, run the command again with --verbose.
make: *** [build] Error 101
```

> I made the changes in src/core/src/signature.rs, line 355 / method `add_protein`, and got:
> 
> ```
>    Compiling sourmash v0.9.0 (/Users/t/dev/sourmash/src/core)
> error[E0277]: the trait bound `(): std::ops::Try` is not satisfied
>    --> src/core/src/signature.rs:366:18
>     |
> 366 |                 .try_for_each(|sketch| {
>     |                  ^^^^^^^^^^^^ the trait `std::ops::Try` is not implemented for `()`
> 
> error: aborting due to previous error
> 
> For more information about this error, try `rustc --explain E0277`.
> error: could not compile `sourmash`.
> 
> To learn more, run the command again with --verbose.
> make: *** [build] Error 101
> ```

Check to see if there is a `;` at the end of that `.try_for_each` call. It should be
```rust
self.signatures 
    .iter_mut()  
    .try_for_each(|sketch| { 
        sketch.add_protein(&seq) }  
    )?;
```
not
```rust
self.signatures 
    .iter_mut()  
    .try_for_each(|sketch| { 
        sketch.add_protein(&seq); }  
    )?;
```
(putting the semicolon makes the `|sketch| { sketch.add_protein(&seq); }` closure return `()`, without the semicolon it is returning `Result<(), Error>` and then  .try_for_each` will work)

--- a/src/core/src/signature.rs
+++ b/src/core/src/signature.rs
@@ -357,15 +357,15 @@ impl Signature {
         if #[cfg(feature = "parallel")] {
             self.signatures
                 .par_iter_mut()
-                .for_each(|sketch| {
-                    sketch.add_protein(&seq).unwrap(); }
-                );
+                .try_for_each(|sketch| {
+                    sketch.add_protein(&seq) }
+                )?;
         } else {
             self.signatures
                 .iter_mut()
-                .for_each(|sketch| {
-                    sketch.add_protein(&seq).unwrap(); }
-                );
+                .try_for_each(|sketch| {
+                    sketch.add_protein(&seq) }
+                )?;
         }
         }
 

--- a/tests/test_sourmash_sketch.py
+++ b/tests/test_sourmash_sketch.py
@@ -9,6 +9,7 @@ import glob
 import json
 import csv
 import pytest
+import screed
 
 from . import sourmash_tst_utils as utils
 import sourmash
@@ -97,6 +98,25 @@ def test_protein_override_bad_2():
     with pytest.raises(ValueError):
         factory = _signatures_for_sketch_factory(['k=21,dna'],
                                                  'protein', False)
+
+def test_protein_override_bad_rust_foo():
+    # mimic 'sourmash sketch protein -p dna'
+    factory = _signatures_for_sketch_factory([], 'protein', False)
+
+    # reach in and avoid error checking to construct a bad params_list.
+    factory.params_list = [('dna', {})]
+
+    # now, get sigs...
+    siglist = factory()
+    assert len(siglist) == 1
+    sig = siglist[0]
+
+    # try adding something
+    testdata1 = utils.get_test_data('ecoli.faa')
+    record = next(iter(screed.open(testdata1)))
+
+    sig.add_protein(record.sequence)
+
 
 def test_dayhoff_defaults():
     factory = _signatures_for_sketch_factory([], 'dayhoff', True)

--- a/tests/test_sourmash_sketch.py
+++ b/tests/test_sourmash_sketch.py
@@ -115,7 +115,10 @@ def test_protein_override_bad_rust_foo():
     testdata1 = utils.get_test_data('ecoli.faa')
     record = next(iter(screed.open(testdata1)))
 
-    sig.add_protein(record.sequence)
+    with pytest.raises(ValueError) as exc:
+        sig.add_protein(record.sequence)
+
+    assert 'Invalid hash function: "dna"' in str(exc)
 
 
 def test_dayhoff_defaults():

